### PR TITLE
Creates a new sessionToken when updating password

### LINKED ISSUE #2252: Cannot create a username/password user from an anonymous user
Hi,

I am running the latest version of the parse-server on my local server. My client is an iOS app.
I create an anonymous user with the following code: PFAnonymousUtils.logInWithBlock

At a certain point in my app, I want to convert this anonymous user to a normal user. I can convert it to a Facebook, Twitter or username/password user.
With FaceBook or Twitter, I use their "linkUser" method and everything works fine.
With the username/password, I do the following:

        var user = PFUser.currenUser()
        user.username   = username?.lowercaseString
        user.email      = username?.lowercaseString
        user.password   = password
        user.signUpInBackgroundWithBlock { (sucess, error) in
        }

When I look at the "Session" table in mongoDB, there is no session created. In fact, the session that was created for the anonymous user was deleted when I tried to transform the anonymous user to a username/password user.
If I log out and log in again, the session is properly created.

Is there a way to link the anonymous user to a username/password user?

### Environment Setup

- **Server**
  - parse-server version: 2.2.16
  - Operating System:     Mac OS X El Capitan
  - Hardware:            MacBook Pro
  - Localhost or remote server? (AWS, Heroku, Azure, Digital Ocean, etc):  localhost

- **Database**
  - MongoDB version: 3.2.1
  - Storage engine:  [FILL THIS OUT]
  - Hardware:        [FILL THIS OUT]
  - Localhost or remote server? (AWS, mLab, ObjectRocket, Digital Ocean, etc): [FILL THIS OUT]

### Logs/Trace

You can turn on additional logging by configuring VERBOSE=1 in your environment.

[Error]: invalid session token (Code: 209, Version: 1.13.0)

### PR DESCRIPTION
**Pull request by flovilmart**
When upgrading a user from anonymous to username/password the original implementation would force the client to call login in order to generate a session token.

This PR attempts to fix that problem by generating a new session token after clearing the sessions.

Fixes #2252

### CODE REVIEW

File: spec/ParseUser.spec.js [hunk 1]
```diff
@@ -2305,7 +2309,30 @@ describe('Parse.User testing', () => {
     }).then((user) => {
       expect(typeof user).toEqual('object');
       expect(user.authData).toBeUndefined();
-      done();
+      expect(user.sessionToken).not.toBeUndefined();
+      // Session token should have changed
+      expect(user.sessionToken).not.toEqual(originalSessionToken);
+      // test that the sessionToken is valid
+      return new Promise((resolve, reject) => {
```
Comment (drew-gross) [inline spec/ParseUser.spec.js]: We have `request-promise` now if you want to avoid some boilerplate in later PRs.

File: src/RestWrite.js [hunk 2]
```diff
@@ -464,15 +469,21 @@ RestWrite.prototype.handleFollowup = function() {
...
     };
     delete this.storage['clearSessions'];
-    this.config.database.destroy('_Session', sessionQuery)
+    return this.config.database.destroy('_Session', sessionQuery)
+    .then(this.handleFollowup.bind(this));
+  }
+  
+  if (this.storage && this.storage['generateNewSession']) {
+    delete this.storage['generateNewSession'];
+    return this.createSessionToken()
```
Comment (drew-gross) [inline src/RestWrite.js]: I think this will be buggy if someone changes email address and upgrade from anon to non-anon in the same request.

**Comment by drew-gross** [comment]
Can you add a test that verification email is sent when upgrading to non-anon and setting email? Also, postgres tests are failing, we may need to add another exclude.

**Comment by flovilmart** [comment]
Yup!

File: spec/ParseUser.spec.js (hunk 1 above)
Comment (flovilmart) [inline spec/ParseUser.spec.js]: That's helpful. I'll update

File: src/RestWrite.js (hunk 2 above)
Comment (flovilmart) [inline src/RestWrite.js]: i'll check it out

**Comment by facebook-github-bot-6** [comment]
@flovilmart updated the pull request.

**Comment by facebook-github-bot-8** [comment]
@flovilmart updated the pull request.

File: src/RestWrite.js (hunk 2 above)
Comment (flovilmart) [inline src/RestWrite.js]: That seem to work correctly as is, as we call the `handleFollowup` after. 

I needed to refactor as synchronous methods so destroy don't race with generation

Comment (drew-gross) [inline src/RestWrite.js]: Oh I see it's a recursive call. Interesting. Yeah, this should work then.

File: spec/ParseUser.spec.js [hunk 3]
```diff
@@ -2305,8 +2309,84 @@ describe('Parse.User testing', () => {
...
+      });
+    }).catch((err) => {
+      fail('no request should fail: ' + JSON.stringify(err));
+      done();
+    });
+  });
+
+  it_exclude_dbs(['postgres'])('should send email when upgrading from anon', (done) => {
+    
+    let emailCalled = false;
```
Comment (drew-gross) [inline spec/ParseUser.spec.js]: You might find jasmine spies more convenient. This is fine here though.

**Pull request merged**